File: services/Bot/src/Crawler.ts
import puppeteer, { Browser, Page } from "puppeteer";
import remove_duplicate from "./utils/remove_duplicate";

const remove_hash_url = (link: string) => {
  if (!link.includes("#")) return link;
  const hash_index = link.indexOf("#");
  return link.substring(0, hash_index);
};

class Scraper {
  private link: string;
  constructor() {
    this.link = "";
  }
  async launch_browser(): Promise<Browser | null> {
    try {
      const browser = await puppeteer.launch();
      return browser;
    } catch (err) {
      const error = err as Error;
      console.log("LOG: Browser closed unexpectedly.");
      console.error(error.message);
      process.exit(1);
    }
  }

  set(new_link: string) {
    this.link = new_link;
    return new_link;
  }
}
type data_t = {
  webpages: Array<{
    header: { title: string; webpage_url: string };
    contents: string;
  }>;
  header: {
    title: string;
    url: string;
  };
};

class Crawler {
  private scraper: Scraper;
  private visited_stack: Set<string>;
  private browser: Browser | null;
  private page: Page | null;
  private max_interval: number = 5000;
  private min_interval: number = 1500;
  private current_interval: number = 500;
  data: data_t;

  constructor(scraper: Scraper) {
    this.scraper = scraper;
    this.visited_stack = new Set<string>([]);
    this.browser = null;
    this.page = null;
    this.data = {
      header: {
        title: "",
        url: "",
      },
      webpages: [],
    };
  }
  async start_crawl(link: string) {
    try {
      this.scraper.set(link);
      this.browser = await this.scraper.launch_browser();
      if (this.browser == null)
        throw new Error("Unable to create browser page.");
      this.page = await this.browser.newPage();
      if (this.page === null) throw new Error("Unable to create browser page.");
      await this.crawl(link);
    } catch (err) {
      const error = err as Error;
      this.browser?.close();
      console.log("LOG: Something went wrong with the crawler.");
      console.error(error.message);
    }
  }

  private async crawl(link: string) {
    try {
      if (this.page === null) throw new Error("Unable to create browser page.");
      await this.page.goto(link);
      this.data.header = { title: await this.page.title(), url: link };
      await this.traverse_pages(link);
      this.browser?.close();
      console.log(`End of Crawl`);
      return this.data;
    } catch (err) {
      const error = err as Error;
      this.browser?.close();
      console.log("LOG: Something went wrong while initializing crawler");
      console.error(error.message);
      return this.data;
    } finally {
      console.log("LOG: Close Browser");
      this.browser?.close();
    }
  }

  private async throttle_visits() {
    const DEC_INC = 70;
    if (
      this.current_interval < this.min_interval ||
      this.current_interval < this.max_interval
    ) {
      this.current_interval += DEC_INC;
    } else if (this.current_interval > this.max_interval) {
      this.current_interval = 500;
    }
    console.log({ current_interval: this.current_interval });
    return new Promise((resolve) => {
      setTimeout(async () => {
        resolve("GO NEXT");
      }, this.current_interval);
    });
  }

  private async traverse_pages(link: string) {
    const current_page = remove_hash_url(link);
    try {
      if (this.browser == null)
        throw new Error("Unable to create browser page.");
      if (this.page == null) throw new Error("Unable to create browser page.");

      if (this.visited_stack.has(current_page)) {
        console.log("Already Visited: " + current_page);
        return this.page;
      }
      this.visited_stack.add(current_page);
      await this.throttle_visits();
      await this.page.goto(current_page);
      await this.index_page(this.page, current_page);
      const css_selector =
        'a:not([href$=".zip"]):not([href$=".pdf"]):not([href$=".exe"]):not([href$=".jpg"]):not([href$=".png"]):not([href$=".tar.gz"]):not([href$=".rar"]):not([href$=".7z"]):not([href$=".mp3"]):not([href$=".mp4"]):not([href$=".mkv"]):not([href$=".tar"]):not([href$=".xz"]):not([href$=".msi"])';
      const extracted_links = await this.page.$$eval(css_selector, (links) =>
        links.map((link) => link.href),
      );
      const neighbors = remove_duplicate<string>(extracted_links).filter(
        (link) => {
          return link.includes(new URL(current_page).origin) ?? link;
        },
      );
      if (neighbors === undefined || neighbors.length === 0) {
        console.log("LOG: No neighbors.");
        return;
      }
      console.log({
        visited_stack: this.visited_stack.size,
        current_trace: current_page,
        current_neighbors: neighbors.length,
      });
      for (let current_neighbor of neighbors) {
        await this.traverse_pages(current_neighbor);
      }
    } catch (err) {
      const error = err as Error;
      this.browser?.close();
      console.log("LOG: Something went wrong when crawling the website");
      console.error(error.message);
    }
  }
  private async index_page(current_page: Page, link: string) {
    if (link.includes("#")) {
      const webpage_from_ds = this.data.webpages.find(
        (el: { header: { title: string; webpage_url: string } }) =>
          remove_hash_url(el.header.webpage_url) === remove_hash_url(link),
      );
      const is_duplicate = webpage_from_ds !== null;
      if (is_duplicate) {
        console.log("DUPLICATES");
        console.log({
          webpage_from_ds: webpage_from_ds!.header.webpage_url,
          link,
        });
        return;
      }
    }

    console.log("NOT DUPLICATES");
    console.log({ link });

    const extract = async () => {
      const filter_data = async (selector: string) => {
        const map_ = await current_page.$$eval(selector, (el) =>
          el.map((p) => {
            return p.textContent;
          }),
        );
        const filtered_ = map_.filter((p) => {
          if (p !== undefined) {
            return p;
          }
        });
        return filtered_.join(" ");
      };
      const paras = await filter_data("p");
      const h1 = await filter_data("h1");
      const h2 = await filter_data("h2");
      const h3 = await filter_data("h3");
      const h4 = await filter_data("h4");
      const code = await filter_data("code");
      const pre = await filter_data("pre");
      return await Promise.allSettled([paras, h1, h2, h3, h4, code, pre]);
    };
    const aggregate_data = async (
      extracted_data: PromiseSettledResult<string>[],
    ) => {
      const settle = extracted_data.map((promises) => {
        if (promises.status === "fulfilled") return promises.value;
        return "";
      });
      const l = settle.join("");
      this.data = {
        ...this.data,
        webpages: [
          ...this.data.webpages,
          {
            header: {
              title: (await current_page.title()) || "",
              webpage_url: current_page.url() || "",
            },
            contents: l,
          },
        ],
      };
    };
    const ex = await extract();
    const ag = await aggregate_data(ex);
  }
}

export { Scraper, Crawler };

File: web-crawler-service-golang/crawler.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	amqp "github.com/rabbitmq/amqp091-go"
	"github.com/tebeka/selenium"
	"log"
	"strings"
	"sync"
	rabbitmqclient "web-crawler-service-golang/pkg/rabbitmq_client"
	webdriver "web-crawler-service-golang/pkg/webdriver"
	utilities "web-crawler-service-golang/utilities/links"
)

type Header struct {
	Title string
	Url   string
}

type WebpageEntry struct {
	URL             string
	IndexedWebpages []IndexedWebpage
	hostname        string
	Title           string
}
type IndexedWebpage struct {
	Header   Header
	Contents string
}

type Crawler struct {
	URL string
	ctx context.Context
	wd  *selenium.WebDriver
}

var elementSelector = []string{
	"a",
	"p",
	"span",
	"pre",
	"h1",
	"h2",
	"h3",
	"h4",
}

type Results struct {
	URLCount    int
	URLsFailed  []string
	Message     string
	ThreadsUsed int
	PageResult  <-chan PageResult
}

var indexedList map[string]IndexedWebpage

const (
	maxRetries   = 7
	threadPool   = 10
	crawlFail    = 0
	crawlSuccess = 1
	// removes links to web objects that does not return an html page.
	linkFilter = `a:not([href$=".zip"]):not([href$=".svg"]):not([href$=".scss"]):not([href$=".css"]):not([href$=".pdf"]):not([href$=".exe"]):not([href$=".jpg"]):not([href$=".png"]):not([href$=".tar.gz"]):not([href$=".rar"]):not([href$=".7z"]):not([href$=".mp3"]):not([href$=".mp4"]):not([href$=".mkv"]):not([href$=".tar"]):not([href$=".xz"]):not([href$=".msi"])`
)

type PageNavigator struct {
	entry        *WebpageEntry
	wd           *selenium.WebDriver
	pagesVisited map[string]string
	currentUrl   string
}

type PageResult struct {
	URL         string
	Message     string
	CrawlStatus int
	TotalPages  int
}

type Message struct {
	Webpages []IndexedWebpage
	Header
}

type ErrorMessage struct {
	Message string
	Url     string
}

type Spawner struct {
	threadPool int
	URLs       []string
}
type Spawn struct {
	threadSlot           chan struct{}
	wg                   *sync.WaitGroup
	ctx                  context.Context
	aggregateResultsChan chan PageResult
}

type ThreadToken struct{}

func (sp *Spawn) SpawnCrawler(doc string) {
	defer func() {
		<-sp.threadSlot
		log.Printf("NOTIF: Thread Token release\n")
	}()
	defer sp.wg.Done()
	wd, err := webdriver.CreateClient()
	if err != nil {
		log.Print(err.Error())
		log.Printf("ERROR: Unable to create a new connection with Chrome Web Driver.\n")
		return
	}
	crawler := Crawler{ctx: sp.ctx, URL: doc, wd: wd}
	result, err := crawler.Crawl()
	if err != nil {
		log.Print(err.Error())
		log.Printf("NOTIF: Thread token release due to error.\n")
		return
	}
	sp.aggregateResultsChan <- result
}

func (s *Spawner) SpawnCrawlers() Results {
	aggregateResultsChan := make(chan PageResult, len(s.URLs))
	threadSlot := make(chan struct{}, s.threadPool)

	var (
		wg  sync.WaitGroup
		ctx context.Context
	)

	wg.Add(1)
	go func() {
		defer wg.Done()
		for _, doc := range s.URLs {
			wg.Add(1)
			threadSlot <- ThreadToken{}
			log.Printf("NOTIF: Thread token insert\n")
			spawn := Spawn{aggregateResultsChan: aggregateResultsChan, threadSlot: threadSlot, wg: &wg, ctx: ctx}
			go spawn.SpawnCrawler(doc)
		}
	}()

	log.Printf("NOTIF: Wait for crawlers\n")
	wg.Wait()
	close(aggregateResultsChan)
	for crawler := range aggregateResultsChan {
		log.Printf("Crawled URL: %s\n", crawler.URL)
		log.Printf("Crawl Message: %s\n", crawler.Message)
	}

	log.Println("NOTIF: All Process have finished.")

	return Results{
		Message:     "Crawled and indexed webpages",
		ThreadsUsed: threadPool,
		URLCount:    len(s.URLs),
		PageResult:  aggregateResultsChan,
	}
}

func (c Crawler) Crawl() (PageResult, error) {
	defer log.Printf("NOTIF: Finished Crawling\n")
	defer (*c.wd).Close()

	log.Printf("NOTIF: Start Crawling %s\n", c.URL)
	hostname, err := utilities.GetHostname(c.URL)
	if err != nil {
		fmt.Println(err.Error())
	}
	entry := WebpageEntry{
		URL:             c.URL,
		IndexedWebpages: make([]IndexedWebpage, 0, 10),
		hostname:        hostname,
		Title:           "",
	}
	pageNavigator := PageNavigator{
		entry:        &entry,
		currentUrl:   c.URL,
		wd:           c.wd,
		pagesVisited: map[string]string{},
	}
	err = pageNavigator.navigatePageWithRetries(maxRetries)
	if err != nil {
		sendErrorOnWebpageCrawl(c.URL)
		fmt.Println(err.Error())
		return PageResult{}, fmt.Errorf("ERROR: Unable to navigate to the website entry point.")
	}
	title, err := (*c.wd).Title()
	if err == nil {
		entry.Title = title
	}
	err = pageNavigator.navigatePages()

	/*
	 TODO when other pages are already indexed, but only a single page throws an error
	 then all progress with huge amounts of data will be lost, need to save these Results
	 despite an error occurs instead of returning an zero byte result.
	*/
	if err != nil {
		sendErrorOnWebpageCrawl(c.URL)
		fmt.Println("ERROR: Well something went wrong with the last stack.")
		return PageResult{
			URL:         c.URL,
			CrawlStatus: crawlFail,
			Message:     "An Error has occured while crawling the current url.",
			TotalPages:  len(entry.IndexedWebpages),
		}, nil
	}
	result := PageResult{
		URL:         c.URL,
		CrawlStatus: crawlSuccess,
		Message:     "Successfully Crawled & Indexed website",
		TotalPages:  len(entry.IndexedWebpages),
	}
	// saveIndexedWebpages(hostname, &entry)
	return result, nil
}

func (pn *PageNavigator) navigatePageWithRetries(retries int) error {
	if retries > 0 {
		err := (*pn.wd).Get(pn.currentUrl)
		if err != nil {
			return pn.navigatePageWithRetries(retries - 1)
		}
		return nil
	}
	return fmt.Errorf("ERROR: Unable to retrieve webpage after several retries.")
}

func (pn *PageNavigator) navigatePages() error {

	if _, visited := pn.pagesVisited[pn.currentUrl]; visited {
		// its so that we can grab unique links and append to children of the current page
		fmt.Println("NOTIF: Page already visited")
		return nil
	}
	err := pn.navigatePageWithRetries(maxRetries)
	if err != nil {
		return fmt.Errorf("ERROR: Unable to visit the current link.")
	}
	pn.pagesVisited[pn.currentUrl] = pn.currentUrl

	links, err := (*pn.wd).FindElements(selenium.ByCSSSelector, linkFilter)

	// no children/error
	if err != nil {
		log.Println("ERROR: Unable to find elements of type `a`.")
		return fmt.Errorf("ERROR: Unable to find elements of type `a`.")
	}

	/*
		Need to check such that we can ignore the already visited links
		and use the ones that doesnt exist and consider it
		as the child of the currently visited link
	*/

	children := make([]string, 0)
	for _, link := range links {
		// need to filter out links that is not the same as hostname
		ref, _ := link.GetAttribute("href")
		cleanedRef, _, _ := strings.Cut(ref, "#")
		childHostname, err := utilities.GetHostname(cleanedRef)

		if err != nil {
			fmt.Println(err.Error())
			continue
		}
		if _, visited := pn.pagesVisited[cleanedRef]; !visited && pn.entry.hostname == childHostname {
			// if so that we can grab unique links and append to children of the current page
			// and ignore links not relative to the entry point link
			children = append(children, cleanedRef)
		}
	}

	indexedWebpage, err := pn.Index()
	if err != nil {
		log.Println("ERROR: Not handled yet")
	}
	pn.entry.IndexedWebpages = append(pn.entry.IndexedWebpages, indexedWebpage)

	/*
	 no child to traverse to then return to caller, the caller function will
	 then go to its next child in the children array.
	*/

	if len(children) == 0 {
		return nil
	}
	for _, child := range children {
		pn.currentUrl = child
		err := pn.navigatePages()
		// if error occured from traversing or any error has occured
		// just move to the next child
		if err != nil {
			continue
		}

	}
	return nil
}
func (pt PageNavigator) Index() (IndexedWebpage, error) {

	/*
		Iterating through the indexSelector, where each selector, we creates
		a new go routine, so using a buffered channel with the exact length of
		the indexSelector would make more sense.

		If ever we want to throttle the operation we can create a semaphore by
		limiting the buffered channel, if resource is an issue.
	*/

	textContentChan := make(chan string, len(elementSelector))
	var wg sync.WaitGroup

	// Start wait group after go routine is processed on a different thread
	wg.Add(1)

	// Go routine generator
	go func() {
		defer func() {
			wg.Done()
			log.Println("NOTIF: Text Extracted from elements.")
		}()
		for _, selector := range elementSelector {
			wg.Add(1)
			go func(selector string) {
				defer wg.Done()
				textContents, err := textContentByElementSelector(pt.wd, selector)
				if err != nil {
					textContentChan <- ""
					return
				}
				textContentChan <- joinContents(textContents)
			}(selector)
		}
	}()

	fmt.Println("NOTIF: Waiting for page indexer")
	wg.Wait()
	close(textContentChan)
	textChanSlice := make([]string, 0, 100)

	/*
		create a select statement to catch an error returned by
		invdividual go routine element selectors
	*/

	for elementContents := range textContentChan {
		textChanSlice = append(textChanSlice, elementContents)
	}

	pageContents := joinContents(textChanSlice)
	title, err := (*pt.wd).Title()
	if err != nil {
		log.Printf("ERROR: No title for this page")
	}

	url, err := (*pt.wd).CurrentURL()
	if err != nil {
		log.Printf("ERROR: No url for this page")
	}
	fmt.Printf("NOTIF: Page %s Indexed\n", url)

	newIndexedPage := IndexedWebpage{
		Contents: pageContents,
		Header: Header{
			Url:   url,
			Title: title,
		},
	}
	return newIndexedPage, nil
}

func textContentByElementSelector(wd *selenium.WebDriver, selector string) ([]string, error) {
	elementTextContents := make([]string, 0, 10)
	elements, err := (*wd).FindElements(selenium.ByCSSSelector, selector)
	if err != nil {
		log.Printf("ERROR: Elements does not satisfy css selector: %s", selector)
		return nil, err
	}
	for _, el := range elements {
		text, err := el.Text()
		if err != nil {
			continue
		}
		elementTextContents = append(elementTextContents, text)
	}

	return elementTextContents, nil
}

func joinContents(tc []string) string {
	return strings.Join(tc, " ")
}

func sendErrorOnWebpageCrawl(hostname string) error {

	/*
	 expressErrorChannel is used to pass a message from crawl service
	 to the express websocket server using the `crawl_poll_queue` (Might change routing key name)
	 routing key to notify users immediately that an error occured for the current
	 to notify users immediately that an error occured for the current url
	*/

	fmt.Println("ERROR Crawl")
	conn, err := rabbitmqclient.GetConnection("receiverConn")
	if err != nil {
		fmt.Print(err.Error())
		log.Panicln("ERROR: Unable to get connection.")
	}
	expressErrorChannel, err := conn.Channel()
	if err != nil {
		log.Printf("ERROR: Unable to create a database channel.")
		return fmt.Errorf(err.Error())
	}

	errorMessage := ErrorMessage{
		Message: "Error",
		Url:     hostname,
	}
	const crawlPollqueue = "crawl_poll_queue"

	expressErrorChannel.QueueDeclare(crawlPollqueue, false, false, false, false, nil)
	dataBuffer, err := json.Marshal(errorMessage)
	expressErrorChannel.Publish("", crawlPollqueue, false, false, amqp.Publishing{
		Type:    "text/plain",
		Body:    []byte(dataBuffer),
		ReplyTo: crawlPollqueue,
	})
	expressErrorChannel.Close()
	return nil
}

func saveIndexedWebpages(jobID string, entry *WebpageEntry) error {
	conn, err := rabbitmqclient.GetConnection("receiverConn")
	if err != nil {
		fmt.Print(err.Error())
		log.Panicln("ERROR: Unable to get connection.")
	}
	dbChannel, err := conn.Channel()
	if err != nil {
		log.Printf("ERROR: Unable to create a database channel.")
		return fmt.Errorf(err.Error())
	}

	resultMessage := Message{
		Webpages: entry.IndexedWebpages,
		Header: Header{
			Title: entry.Title,
			Url:   entry.hostname,
		},
	}
	const (
		dbIndexingQueue = "db_indexing_crawler"
		crawlPollqueue  = "crawl_poll_queue"
	)
	dbChannel.QueueDeclare(dbIndexingQueue, false, false, false, false, nil)
	dataBuffer, err := json.Marshal(resultMessage)
	dbChannel.Publish("", dbIndexingQueue, false, false, amqp.Publishing{
		Type:          "text/plain",
		Body:          []byte(dataBuffer),
		CorrelationId: jobID,
		ReplyTo:       crawlPollqueue,
	})
	dbChannel.Close()
	return nil
}
